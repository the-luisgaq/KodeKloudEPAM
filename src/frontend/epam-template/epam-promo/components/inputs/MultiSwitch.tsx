import { createSkinComponent } from '@epam/uui-core';
import * as uui from '@epam/uui';

type MultiSwitchMods = {
    /**
     * Defines component color.
     * @default 'blue'
     */
    color?: 'blue' | 'gray50' | 'gray' | uui.MultiSwitchProps['color'];

    /**
     * Defines component size.
     * @default '36'
     */
    size?: '24' | '30' | '36' | '42' | '48' | '60';
};

/** Represents the properties for the MultiSwitch component. */
export type MultiSwitchProps<TValue> = uui.MultiSwitchCoreProps<TValue> & MultiSwitchMods;

export const MultiSwitch = createSkinComponent<uui.MultiSwitchProps<any>, MultiSwitchProps<any>>(
    uui.MultiSwitch,
    (props) => {
        return {
            color: props.color || 'blue',
        };
    },
);
