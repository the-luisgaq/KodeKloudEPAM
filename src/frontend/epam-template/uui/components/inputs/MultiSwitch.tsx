import * as React from 'react';
import { IEditable, IHasRawProps, Overwrite } from '@epam/uui-core';
import { ControlGroup } from '../layout/ControlGroup';
import { Button, ButtonProps } from '../buttons';

import type { JSX } from 'react';

type MultiSwitchItem = ButtonProps & {
    /**
     * Defines the id of MultiSwitchItem.
     */
    id: any;
};

interface MultiSwitchMods {
    /**
     * Defines component color.
     * @default 'primary'
     */
    color?: 'primary' | 'secondary';
    /**
     * Defines component size.
     * @default '36'
     */
    size?: '24' | '30' | '36' | '42' | '48' | '60';
}

export interface MultiSwitchModsOverride {}

/** Represents the 'Core properties' for the MultiSwitch component. */
export type MultiSwitchCoreProps<TValue> = IEditable<TValue> & IHasRawProps<React.HTMLAttributes<HTMLDivElement>> & {
    /**
     * Defines an array of MultiSwitchItems.
     */
    items: MultiSwitchItem[];
};

/** Represents the properties for the MultiSwitch component. */
export type MultiSwitchProps<TValue = unknown> = MultiSwitchCoreProps<TValue> & Overwrite<MultiSwitchMods, MultiSwitchModsOverride>;

function MultiSwitchComponent<TValue>(props: MultiSwitchProps<TValue>, ref: React.ForwardedRef<HTMLDivElement>) {
    return (
        <ControlGroup
            ref={ ref }
            rawProps={ {
                ...props.rawProps,
                role: 'tablist',
                'aria-invalid': props.isInvalid,
                'aria-required': props.isRequired,
                'aria-disabled': props.isDisabled,
                'aria-readonly': props.isReadonly,
            } }
        >
            {props.items.map((item, index) => (
                <Button
                    { ...props }
                    { ...item }
                    isDisabled={ props.isDisabled }
                    key={ index + '-' + item.id }
                    onClick={ !props.isReadonly && (() => props.onValueChange(item.id)) }
                    fill={ props.value === item.id ? 'solid' : 'outline' }
                    color={ props.color }
                    size={ props.size }
                    rawProps={ { 'aria-current': props.value === item.id, role: 'tab' } }
                />
            ))}
        </ControlGroup>
    );
}

export const MultiSwitch = React.forwardRef(MultiSwitchComponent) as <TValue>(props: MultiSwitchProps<TValue>) => JSX.Element;
