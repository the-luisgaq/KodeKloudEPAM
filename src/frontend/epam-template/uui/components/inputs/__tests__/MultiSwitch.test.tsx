import React from 'react';
import { MultiSwitch } from '../MultiSwitch';
import { renderSnapshotWithContextAsync } from '@epam/uui-test-utils';

it('should be rendered correctly', async () => {
    const tree = await renderSnapshotWithContextAsync(
        <MultiSwitch
            value={ 1 }
            onValueChange={ jest.fn }
            items={ [{ id: 1, caption: 'On' }, { id: 2, caption: 'Off' }] }
        />,
    );
    expect(tree).toMatchSnapshot();
});
