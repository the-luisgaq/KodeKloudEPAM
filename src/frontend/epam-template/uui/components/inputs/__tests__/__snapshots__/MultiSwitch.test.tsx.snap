// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be rendered correctly 1`] = `
<DocumentFragment>
  <div
    class="container root uui-control-group"
    role="tablist"
  >
    <button
      aria-current="true"
      class="uui-button-box uui-enabled -clickable container root uui-button uui-fill-solid uui-color-primary uui-size-36"
      role="tab"
      tabindex="0"
      type="button"
    >
      <div
        class="uui-caption"
      >
        On
      </div>
    </button>
    <button
      aria-current="false"
      class="uui-button-box uui-enabled -clickable container root uui-button uui-fill-outline uui-color-primary uui-size-36"
      role="tab"
      tabindex="0"
      type="button"
    >
      <div
        class="uui-caption"
      >
        Off
      </div>
    </button>
  </div>
</DocumentFragment>
`;
